Zend/zend_compile.c

```diff
@@ -7522,9 +7572,81 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
 			opline->result.opline_num = -1;
 		}
 	}
+
+	return ce;
 }
 /* }}} */
 
+static void zend_compile_enum_case(zend_ast *ast)
+{
+	zend_class_entry *enum_class = CG(active_class_entry);
+	if (!(enum_class->ce_flags & ZEND_ACC_ENUM)) {
+		zend_error_noreturn(E_COMPILE_ERROR, "Case can only be used in enums");
+	}
+
+	zend_string *enum_case_name = zend_ast_get_str(ast->child[0]);
+	zend_string *enum_class_name = enum_class->name;
+
+	zval class_name_zval;
+	ZVAL_STR_COPY(&class_name_zval, enum_class_name);
+	zend_ast *class_name_ast = zend_ast_create_zval_ex(&class_name_zval, ZEND_NAME_NOT_FQ);
+
+	zval case_name_zval;
+	ZVAL_STR_COPY(&case_name_zval, enum_case_name);
```

Comment:
`make test TESTS=Zend/tests/enum/scalar-negative-int.phpt\ -d\ opcache.enable_cli=1\ -d\ opcache.enable=1\ --show-diff\ -m\ --show-mem` fails - I wonder if this needs to create a new interned string and use that instead before inserting enum_case_name. Not familiar with the interning code.

```
==110459== Invalid read of size 8
==110459==    at 0x563CEB: zend_string_hash_val (zend_string.h:106)
==110459==    by 0x563CEB: zend_hash_find_bucket (zend_hash.c:649)
==110459==    by 0x563CEB: zend_hash_find (zend_hash.c:2246)
==110459==    by 0x5DE4B1: zend_enum_from_func (zend_enum.c:253)
==110459==    by 0x5B897F: ZEND_DO_FCALL_SPEC_RETVAL_USED_HANDLER (zend_vm_execute.h:1863)
==110459==    by 0x5B897F: execute_ex (zend_vm_execute.h:54549)
==110459==    by 0x5B9BE3: zend_execute (zend_vm_execute.h:58868)
==110459==    by 0x551332: zend_execute_scripts (zend.c:1675)
==110459==    by 0x4ED861: php_execute_script (main.c:2488)
==110459==    by 0x5E0604: do_cli (php_cli.c:963)
==110459==    by 0x33CB61: main (php_cli.c:1356)
==110459==  Address 0x5230508 is 8 bytes inside a block of size 32 free'd
==110459==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==110459==    by 0x53B81C4: zend_string_release (zend_string.h:313)
==110459==    by 0x53B81C4: accel_new_interned_string (ZendAccelerator.c:513)
==110459==    by 0x53C9CEE: zend_persist_zval_calc (zend_persist_calc.c:111)
==110459==    by 0x53C9E2F: zend_persist_ast_calc (zend_persist_calc.c:99)
==110459==    by 0x53CB05C: zend_persist_class_constant_calc (zend_persist_calc.c:357)
==110459==    by 0x53CB05C: zend_persist_class_entry_calc (zend_persist_calc.c:440)
==110459==    by 0x53CB05C: zend_accel_persist_class_table_calc (zend_persist_calc.c:541)
==110459==    by 0x53CB05C: zend_accel_script_persist_calc (zend_persist_calc.c:585)
==110459==    by 0x53C1214: cache_script_in_shared_memory (ZendAccelerator.c:1487)
==110459==    by 0x53C1214: persistent_compile_file (ZendAccelerator.c:2176)
==110459==    by 0x53C1214: persistent_compile_file (ZendAccelerator.c:1956)
==110459==    by 0x5512FC: zend_execute_scripts (zend.c:1669)
==110459==    by 0x4ED861: php_execute_script (main.c:2488)
==110459==    by 0x5E0604: do_cli (php_cli.c:963)
==110459==    by 0x33CB61: main (php_cli.c:1356)
==110459==  Block was alloc'd at
==110459==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==110459==    by 0x525D3C: __zend_malloc (zend_alloc.c:3030)
==110459==    by 0x5171A1: zend_string_alloc (zend_string.h:141)
==110459==    by 0x5171A1: zend_string_init (zend_string.h:163)
==110459==    by 0x5171A1: lex_scan (zend_language_scanner.l:3086)
==110459==    by 0x52F749: zendlex (zend_compile.c:1796)
==110459==    by 0x50FFE1: zendparse (zend_language_parser.c:4535)
==110459==    by 0x5129C1: zend_compile (zend_language_scanner.l:611)
==110459==    by 0x5142A9: compile_file (zend_language_scanner.l:665)
==110459==    by 0x420831: phar_compile_file (phar.c:3339)
==110459==    by 0x53BE44C: opcache_compile_file.isra.0 (ZendAccelerator.c:1790)
==110459==    by 0x53C1129: persistent_compile_file (ZendAccelerator.c:2167)
==110459==    by 0x53C1129: persistent_compile_file (ZendAccelerator.c:1956)
==110459==    by 0x5512FC: zend_execute_scripts (zend.c:1669)
==110459==    by 0x4ED861: php_execute_script (main.c:2488)
==110459== 
==110459== Invalid read of size 4
==110459==    at 0x563D00: zend_hash_find_bucket (zend_hash.c:653)
==110459==    by 0x563D00: zend_hash_find (zend_hash.c:2246)
==110459==    by 0x5DE4B1: zend_enum_from_func (zend_enum.c:253)
==110459==    by 0x5B897F: ZEND_DO_FCALL_SPEC_RETVAL_USED_HANDLER (zend_vm_execute.h:1863)
==110459==    by 0x5B897F: execute_ex (zend_vm_execute.h:54549)
==110459==    by 0x5B9BE3: zend_execute (zend_vm_execute.h:58868)
==110459==    by 0x551332: zend_execute_scripts (zend.c:1675)
==110459==    by 0x4ED861: php_execute_script (main.c:2488)
==110459==    by 0x5E0604: do_cli (php_cli.c:963)
==110459==    by 0x33CB61: main (php_cli.c:1356)
==110459==  Address 0x52322f8 is 40 bytes inside a block of size 320 free'd
==110459==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==110459==    by 0x53C5D59: zend_hash_persist (zend_persist.c:153)
==110459==    by 0x53C7E8F: zend_persist_class_entry (zend_persist.c:847)
==110459==    by 0x53C7E8F: zend_accel_persist_class_table (zend_persist.c:1147)
==110459==    by 0x53C7E8F: zend_accel_script_persist (zend_persist.c:1211)
==110459==    by 0x53C126C: cache_script_in_shared_memory (ZendAccelerator.c:1545)
==110459==    by 0x53C126C: persistent_compile_file (ZendAccelerator.c:2176)
==110459==    by 0x53C126C: persistent_compile_file (ZendAccelerator.c:1956)
==110459==    by 0x5512FC: zend_execute_scripts (zend.c:1669)
==110459==    by 0x4ED861: php_execute_script (main.c:2488)
==110459==    by 0x5E0604: do_cli (php_cli.c:963)
==110459==    by 0x33CB61: main (php_cli.c:1356)
==110459==  Block was alloc'd at
==110459==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==110459==    by 0x525D3C: __zend_malloc (zend_alloc.c:3030)
==110459==    by 0x55D501: zend_hash_real_init_mixed_ex (zend_hash.c:171)
==110459==    by 0x55D501: zend_hash_real_init_mixed (zend_hash.c:334)
==110459==    by 0x55E4C0: _zend_hash_add_or_update_i (zend_hash.c:735)
==110459==    by 0x55E4C0: zend_hash_add (zend_hash.c:896)
==110459==    by 0x55B271: zend_hash_add_ptr (zend_hash.h:611)
==110459==    by 0x55B271: zend_declare_class_constant_ex (zend_API.c:4107)
==110459==    by 0x53CF84: zend_compile_enum_case (zend_compile.c:7639)
==110459==    by 0x53CF84: zend_compile_stmt (zend_compile.c:9617)
==110459==    by 0x53D9DE: zend_compile_stmt_list (zend_compile.c:6033)
==110459==    by 0x53CBEF: zend_compile_stmt (zend_compile.c:9561)
==110459==    by 0x53E86A: zend_compile_class_decl (zend_compile.c:7483)
==110459==    by 0x540047: zend_compile_top_stmt (zend_compile.c:9536)
==110459==    by 0x540077: zend_compile_top_stmt (zend_compile.c:9525)
==110459==    by 0x512A3E: zend_compile (zend_language_scanner.l:630)
==110459== 
....
```